readium/navigator/src/main/java/org/readium/r2/navigator/settings/Preferences.kt

```diff
@@ -0,0 +1,395 @@
+/*
+ * Copyright 2022 Readium Foundation. All rights reserved.
+ * Use of this source code is governed by the BSD-style license
+ * available in the top-level LICENSE file of the project.
+ */
+
+package org.readium.r2.navigator.settings
+
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.json.JsonElement
+import kotlinx.serialization.json.JsonObject
+import kotlinx.serialization.json.buildJsonObject
+import org.readium.r2.shared.DelicateReadiumApi
+import org.readium.r2.shared.ExperimentalReadiumApi
+import org.readium.r2.shared.extensions.tryOrLog
+
+/**
+ * Set of preferences used to update a [Configurable]'s settings.
+ *
+ * To modify the preferences, use one of the builder methods based on [MutablePreferences]. It
+ * offers convenient helpers to update different type of [Setting] objects.
+ *
+ * [Preferences] can be serialized to JSON, which is useful to persist user preferences.
+ *
+ * Usage example:
+ *
+ *     // Get the currently available settings for the configurable.
+ *     val settings = configurable.settings.value
+ *
+ *     // Build a new set of Preferences, using the Setting objects as keys.
+ *     val prefs = Preferences {
+ *         set(settings.overflow, Overflow.PAGINATED)
+ *         increment(settings.fontSize)
+ *     }
+ *
+ *     // Apply the preferences to the Configurable, which will automatically update its settings
+ *     // accordingly.
+ *     configurable.applyPreferences(prefs)
+ *
+ * @param values Direct access to the JSON values. Prefer using the safe [Setting]-based accessors
+ * instead.
+ */
+@ExperimentalReadiumApi
+@OptIn(DelicateReadiumApi::class)
+open class Preferences(
+    @DelicateReadiumApi open val values: Map<String, JsonElement> = emptyMap()
+)  {
+
+    /**
+     * Creates a [Preferences] object using a mutable builder.
+     */
+    constructor(builder: MutablePreferences.() -> Unit)
+        : this(MutablePreferences().apply(builder))
+
+    /**
+     * Creates a read-only [Preferences] object from a mutable one.
+     */
+    constructor(preferences: MutablePreferences)
+        : this(preferences.values.toMap())
+
+    /**
+     * Creates a [Preferences] object from its JSON representation.
+     */
+    constructor(jsonString: String?)
+        : this(jsonString
```

Comment:
This constructor cannot fail, the `tryOrLog` will catch any exception and just return an empty `Preferences`, since there's not much that can be done in this case.

Reply:
As usual, I would let apps be aware that they're doing bullshit. We can just offer proper parsing tool, and constructors for proper parsed data.